pkg/server/singleprocess/service_ui_pipeline_run.go

```diff
@@ -72,3 +74,96 @@ func (s *Service) UI_ListPipelineRuns(
 		Pagination:         &pb.PaginationResponse{},
 	}, nil
 }
+
+func (s *Service) UI_GetPipelineRun(
+	ctx context.Context,
+	req *pb.UI_GetPipelineRunRequest,
+) (*pb.UI_GetPipelineRunResponse, error) {
+	log := hclog.FromContext(ctx)
+
+	if err := serverptypes.ValidateUIGetPipelineRunRequest(req); err != nil {
+		return nil, err
+	}
+
+	runResp, err := s.GetPipelineRun(ctx, &pb.GetPipelineRunRequest{
+		Pipeline: req.Pipeline,
+		Sequence: req.Sequence,
+	})
+	if err != nil {
+		return nil, err
+	}
+	run := runResp.PipelineRun
+
+	// Fetch full jobs
+	start := time.Now()
+	var jobs []*pb.Job
+	for _, ref := range run.Jobs {
+		job, err := s.GetJob(ctx, &pb.GetJobRequest{JobId: ref.Id})
+		if err != nil {
+			return nil, hcerr.Externalize(
+				log,
+				err,
+				"failed to get jobs for all pipeline steps",
+			)
+		}
+		jobs = append(jobs, job)
+	}
+	metrics.MeasureOperation(ctx, start, "fetch_jobs_for_ui_get_pipeline_run")
+
+	// Fetch latest status report for every deployment and release
```

Comment:
Good idea, I’ll take a look!

Reply:
I gave [`getDeploymentBundle`](https://github.com/hashicorp/waypoint/blob/613b24583d0d361e70ea38e34eda0a09293cd7f6/pkg/server/singleprocess/service_ui_deploy.go#L134) a spin and it wasn’t that much simpler…

```diff
  if d := job.Result.GetDeploy().GetDeployment(); d != nil {
-          sr, err := s.GetLatestStatusReport(ctx, &pb.GetLatestStatusReportRequest{
-                  Application: d.Application,
-                  Workspace:   d.Workspace,
-                  Target: &pb.GetLatestStatusReportRequest_DeploymentId{
-                          DeploymentId: d.Id,
-                  },
-          })
-          if err != nil {
-                  return nil, hcerr.Externalize(
-                          log,
-                          err,
-                          "failed to get latest status report for deployment %q",
-                          d.Id,
-                  )
-          }
+          bundle, err := s.getDeploymentBundle(ctx, d.Application, d.Workspace, d)
+          if err != nil {
+                  return nil, hcerr.Externalize(
+                          hclog.FromContext(ctx),
+                          err,
+                          "error getting deployment bundle for application",
+                          "application",
+                          d.Application.GetApplication(),
+                          "workspace",
+                          d.Workspace.GetWorkspace(),
+                  )
+          }
+          sr := bundle.LatestStatusReport

           if sr != nil {
                   statusReports = append(statusReports, sr)
           }
  }
  ```

Is this what you had in mind, or something different?

Reply:
Yep! I guess in this case we're only caring about the status report since we already have the deployment, so yeah maybe not too useful of an optimization. Worth a shot!